File: services/service-plugin/controllers/notion.go
package controllers

import (
	"net/http"

	"github.com/Samudai/service-plugin/internal/notion"
	notionpkg "github.com/Samudai/service-plugin/pkg/notion"
	"github.com/gin-gonic/gin"
)

// NotionAuthParam is the paramter for NotionAuth
type NotionAuthParam struct {
	Code        string `json:"code" binding:"required"`
	MemberID    string `json:"member_id" binding:"required"`
	RedirectURI string `json:"redirect_uri" binding:"required"`
}

func NotionAuth(c *gin.Context) {
	var params NotionAuthParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	auth, err := notion.Auth(params.MemberID, params.Code, params.RedirectURI)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, auth)
}

func CheckMemberNotionExists(c *gin.Context) {
	memberID := c.Param("member_id")

	exists, username, err := notion.CheckMemberNotionExists(memberID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"exists": exists, "username": username})
}

// GetAllDatabaseParam is the paramter for GetDatabase
type GetAllDatabaseParam struct {
	MemberID string `json:"member_id" binding:"required"`
}

// GetAllDatabase is the handler for GetAllDatabase
func GetAllDatabase(c *gin.Context) {
	var params GetAllDatabaseParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	auth, err := notion.Auth(params.MemberID, "", "")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	data, err := notion.GetAllDatabase(auth.AccessToken)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"databases": data})
}

// GetDatabase is the handler for GetDatabase
func GetDatabase(c *gin.Context) {
	var params notionpkg.GetDatabaseParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	auth, err := notion.Auth(params.MemberID, "", "")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	database, err := notion.GetDatabase(params.DatabaseID, auth.AccessToken)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, *database)
}

func GetDatabaseProperties(c *gin.Context) {
	var params notionpkg.GetDatabaseParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	auth, err := notion.Auth(params.MemberID, "", "")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	properties, err := notion.GetDatabaseProperties(params.DatabaseID, auth.AccessToken)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, properties)
}

func GetPages(c *gin.Context) {
	var params notionpkg.GetDatabaseParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	auth, err := notion.Auth(params.MemberID, "", "")
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	pages, err := notion.QueryDatabase(auth.AccessToken, params.DatabaseID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	err = notion.SavePages(pages)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, pages)
}

type GetMemberIDsParam struct {
	UserIDs []string `json:"user_ids" binding:"required"`
}

func GetNotionMemberIDs(c *gin.Context) {
	var params GetMemberIDsParam
	if err := c.ShouldBindJSON(&params); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	memberIDs, err := notion.GetMemberIDs(params.UserIDs)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, memberIDs)
}

func DeleteNotionAuth(c *gin.Context) {
	memberID := c.Param("member_id")

	err := notion.DeleteAuth(memberID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "success"})
}

// Extras

func GetPageByID(c *gin.Context) {
	pageID := c.Param("page_id")

	page, err := notion.GetPageByID(pageID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"page": page})
}

File: backend/gateway-consumer/src/lib/notion.ts
import axios from 'axios';
import { NotionDatabase } from '@samudai_xyz/gateway-consumer-types';
import { PageObjectResponse } from '../utils/notionApiEndpoint';
import { AuthResponse } from '../utils/types';

const serviceNotion = `${process.env.SERVICE_PLUGIN}/plugins/notion`;

export const auth = async (member_id: string, code: string, redirect_uri: string): Promise<AuthResponse> => {
    try {
        const response = await axios.post(`${serviceNotion}/auth`, {
            member_id,
            code,
            redirect_uri,
        });
        return response.data;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getAllDatabase = async (member_id: string): Promise<NotionDatabase[]> => {
    try {
        const response = await axios.post(`${serviceNotion}/getalldatabase`, {
            member_id,
        });
        return response.data.databases;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getDatabaseProperties = async (member_id: string, database_id: string) => {
    try {
        const response = await axios.post(`${serviceNotion}/getdatabaseproperties`, {
            member_id,
            database_id,
        });
        return response.data;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getDatabase = async (member_id: string, database_id: string) => {
    try {
        const response = await axios.post(`${serviceNotion}/getdatabase`, {
            member_id,
            database_id,
        });
        return response.data;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getPages = async (member_id: string, database_id: string) => {
    try {
        const response = await axios.post(`${serviceNotion}/getpages`, {
            member_id,
            database_id,
        });
        return response.data;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getMemberIDs = async (people: any) => {
    try {
        let user_ids: string[] = [];
        people.map((user: any) => {
            user_ids.push(user.id);
        });

        const response = await axios.post(`${serviceNotion}/getmemberids`, {
            user_ids,
        });
        return response.data;
    } catch (err: any) {
        throw new Error(err);
    }
};

export const getPageByID = async (pageId: string): Promise<PageObjectResponse> => {
    try {
        const response = await axios.get(`${serviceNotion}/page/${pageId}`);
        return response.data.page;
    } catch (err: any) {
        throw new Error(err);
    }
};
